DOC:Remove hard-coded examples from _flex_doc_SERIES (#24589)

danielplawrence opened the pull request:
Initial work on #24589 

- Removes hard-coded examples from _flex_doc_SERIES
- Adds separate examples for each op (_*_example_SERIES). At this stage I've just copied the examples which were in the _flex_doc_SERIES template (add, sub, mul, div)
- Modifies _make_flex_doc to format the _flex_doc_SERIES template with an example string from op_desc['series_examples']
- adds references to each of the examples to _op_descriptions['series_examples'], allowing them to be picked up in _make_flex_doc
 
Ops outside not in [add, sub, mul, div] will return their docstring with no examples in this revision.

danielplawrence committed:
DOC:Remove hard-coded examples from _flex_doc_SERIES (#24589)

TomAugspurger commented on pandas/core/ops.py:
```diff
@@ -384,57 +384,175 @@ def _get_op_name(op, special):
+d    1.0
+e    NaN
+dtype: float64
+>>> a.multiply(b)
```
Maybe provide a `fill_value` here, to be consistent with the others?

TomAugspurger commented:
Thanks for working on this. I think it's OK that other ops like mod (temporarily) lack examples.

Two questions

1. Maybe add a `fill_value` for Series.multiply
2. How to test this? It'd be nice if we could run the doctests, but I don't know how easy that is due to how we define these methods. Do you have any thoughts?

danielplawrence committed:
DOC:Add fill_value to example for SERIES.multiply() (#24589)

danielplawrence committed:
DOC:Add examples for SERIES ops -- mod,pow,truediv,floordiv (#24589)

danielplawrence committed:
DOC:Use string concatenation in _make_flex_doc to appease docstring validator (#24589)

pep8speaks commented:
Hello @danielplawrence! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/ops.py`](https://github.com/pandas-dev/pandas/blob/fc60c5fddd2c515f64ccce60159c3dd9906d5f51/pandas/core/ops.py):

> [Line 1059:80](https://github.com/pandas-dev/pandas/blob/fc60c5fddd2c515f64ccce60159c3dd9906d5f51/pandas/core/ops.py#L1059): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

danielplawrence committed:
DOC:Fix a long line to satisfy PEP8 (#24589)

danielplawrence commented:
Brief update:
- I've gone through and added examples for several other ops. We now have:
    - add
    - sub
    - mul
    - mod
    - pow
    - truediv
    - floordiv
- These are all just examples with the same input arrays of 1s and NaN which were in the previous docstring.
- Added fill_value=0 to each example as suggested above.
- Didn't do divmod as I couldn't get it to work. divmod() works fine, but Series.divmod() fails with a ValueError, see example below.
<details>
<summary>Series.divmod examples</summary>
<pre>
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> divmod(a,b)
(a    1.0
b    NaN
c    NaN
d    NaN
e    NaN
dtype: float64, a    0.0
b    NaN
c    NaN
d    NaN
e    NaN
dtype: float64)
>>> a.divmod(b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/danlaw/Projects/pandas/pandas/core/ops.py", line 2046, in flex_wrapper
    return self._binop(other, op, level=level, fill_value=fill_value)
  File "/Users/danlaw/Projects/pandas/pandas/core/series.py", line 2522, in _binop
    result = self._constructor(result, index=new_index, name=name)
  File "/Users/danlaw/Projects/pandas/pandas/core/series.py", line 250, in __init__
    .format(val=len(data), ind=len(index)))
ValueError: Length of passed values is 2, index implies 5
</pre>
</details>

- Checked the doctests with validate_docstrings.py, these are validating correctly and fail if I introduce incorrect examples. 
- Doctests raised several errors: (full output attached below)
    - Complained about the spaces introduced by using a format string at the end of the template. I've fixed this so that _make_flex_doc() simply concatenates the base doc with the examples -- this removes trailing whitespace from the cases with no examples.
    - These errors:
          **a)	Parameters {axis} not documented**
*It seems that Series ops can take a parameter which raises 'No axis named a for object type <class 'type'>' for values other than 0. Should this parameter be removed from Series? Doesn't seem to make sense to document it unless I'm missing something. I've added an example at the bottom of this comment.*
	  **b)   Parameter "other" has no description**  
*Appears to be a false positive*
	  **c) Missing description for See Also "Series.rpow" reference**
*False positive?*

- I see that Azure piplines threw a linting failure ([link](https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=8995&view=logs)), I'll look into it ASAP

<details>
  <summary>Full output of validate_docstrings.py</summary>
<pre>
(pandas-dev) 8c8590165d94% for op in add sub mul mod pow truediv floordiv
do
python3 scripts/validate_docstrings.py pandas.Series.$op
done

################################################################################
######################## Docstring (pandas.Series.add)  ########################
################################################################################

Return Addition of series and other, element-wise (binary operator `add`).

Equivalent to ``series + other``, but with support to substitute a fill_value for
missing data in one of the inputs.

Parameters
----------
other : Series or scalar value
fill_value : None or float value, default None (NaN)
    Fill existing missing (NaN) values, and any new element needed for
    successful Series alignment, with this value before computation.
    If data in both corresponding Series locations is missing
    the result will be missing.
level : int or name
    Broadcast across a level, matching Index values on the
    passed MultiIndex level.

Returns
-------
Series
    The result of the operation.

See Also
--------
Series.radd

Examples
--------
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> a
a    1.0
b    1.0
c    1.0
d    NaN
dtype: float64
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> b
a    1.0
b    NaN
d    1.0
e    NaN
dtype: float64
>>> a.add(b, fill_value=0)
a    2.0
b    1.0
c    1.0
d    1.0
e    NaN
dtype: float64

################################################################################
################################## Validation ##################################
################################################################################

3 Errors found:
	Parameters {axis} not documented
	Parameter "other" has no description
	Missing description for See Also "Series.radd" reference

################################################################################
######################## Docstring (pandas.Series.sub)  ########################
################################################################################

Return Subtraction of series and other, element-wise (binary operator `sub`).

Equivalent to ``series - other``, but with support to substitute a fill_value for
missing data in one of the inputs.

Parameters
----------
other : Series or scalar value
fill_value : None or float value, default None (NaN)
    Fill existing missing (NaN) values, and any new element needed for
    successful Series alignment, with this value before computation.
    If data in both corresponding Series locations is missing
    the result will be missing.
level : int or name
    Broadcast across a level, matching Index values on the
    passed MultiIndex level.

Returns
-------
Series
    The result of the operation.

See Also
--------
Series.rsub

Examples
--------
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> a
a    1.0
b    1.0
c    1.0
d    NaN
dtype: float64
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> b
a    1.0
b    NaN
d    1.0
e    NaN
dtype: float64
>>> a.subtract(b, fill_value=0)
a    0.0
b    1.0
c    1.0
d   -1.0
e    NaN
dtype: float64

################################################################################
################################## Validation ##################################
################################################################################

3 Errors found:
	Parameters {axis} not documented
	Parameter "other" has no description
	Missing description for See Also "Series.rsub" reference

################################################################################
######################## Docstring (pandas.Series.mul)  ########################
################################################################################

Return Multiplication of series and other, element-wise (binary operator `mul`).

Equivalent to ``series * other``, but with support to substitute a fill_value for
missing data in one of the inputs.

Parameters
----------
other : Series or scalar value
fill_value : None or float value, default None (NaN)
    Fill existing missing (NaN) values, and any new element needed for
    successful Series alignment, with this value before computation.
    If data in both corresponding Series locations is missing
    the result will be missing.
level : int or name
    Broadcast across a level, matching Index values on the
    passed MultiIndex level.

Returns
-------
Series
    The result of the operation.

See Also
--------
Series.rmul

Examples
--------
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> a
a    1.0
b    1.0
c    1.0
d    NaN
dtype: float64
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> b
a    1.0
b    NaN
d    1.0
e    NaN
dtype: float64
>>> a.multiply(b, fill_value=0)
a    1.0
b    0.0
c    0.0
d    0.0
e    NaN
dtype: float64

################################################################################
################################## Validation ##################################
################################################################################

3 Errors found:
	Parameters {axis} not documented
	Parameter "other" has no description
	Missing description for See Also "Series.rmul" reference

################################################################################
######################## Docstring (pandas.Series.mod)  ########################
################################################################################

Return Modulo of series and other, element-wise (binary operator `mod`).

Equivalent to ``series % other``, but with support to substitute a fill_value for
missing data in one of the inputs.

Parameters
----------
other : Series or scalar value
fill_value : None or float value, default None (NaN)
    Fill existing missing (NaN) values, and any new element needed for
    successful Series alignment, with this value before computation.
    If data in both corresponding Series locations is missing
    the result will be missing.
level : int or name
    Broadcast across a level, matching Index values on the
    passed MultiIndex level.

Returns
-------
Series
    The result of the operation.

See Also
--------
Series.rmod

Examples
--------
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> a
a    1.0
b    1.0
c    1.0
d    NaN
dtype: float64
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> b
a    1.0
b    NaN
d    1.0
e    NaN
dtype: float64
>>> a.mod(b, fill_value=0)
a    0.0
b    NaN
c    NaN
d    0.0
e    NaN
dtype: float64

################################################################################
################################## Validation ##################################
################################################################################

3 Errors found:
	Parameters {axis} not documented
	Parameter "other" has no description
	Missing description for See Also "Series.rmod" reference

################################################################################
######################## Docstring (pandas.Series.pow)  ########################
################################################################################

Return Exponential power of series and other, element-wise (binary operator `pow`).

Equivalent to ``series ** other``, but with support to substitute a fill_value for
missing data in one of the inputs.

Parameters
----------
other : Series or scalar value
fill_value : None or float value, default None (NaN)
    Fill existing missing (NaN) values, and any new element needed for
    successful Series alignment, with this value before computation.
    If data in both corresponding Series locations is missing
    the result will be missing.
level : int or name
    Broadcast across a level, matching Index values on the
    passed MultiIndex level.

Returns
-------
Series
    The result of the operation.

See Also
--------
Series.rpow

Examples
--------
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> a
a    1.0
b    1.0
c    1.0
d    NaN
dtype: float64
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> b
a    1.0
b    NaN
d    1.0
e    NaN
dtype: float64
>>> a.pow(b, fill_value=0)
a    1.0
b    1.0
c    1.0
d    0.0
e    NaN
dtype: float64

################################################################################
################################## Validation ##################################
################################################################################

3 Errors found:
	Parameters {axis} not documented
	Parameter "other" has no description
	Missing description for See Also "Series.rpow" reference

################################################################################
###################### Docstring (pandas.Series.truediv)  ######################
################################################################################

Return Floating division of series and other, element-wise (binary operator `truediv`).

Equivalent to ``series / other``, but with support to substitute a fill_value for
missing data in one of the inputs.

Parameters
----------
other : Series or scalar value
fill_value : None or float value, default None (NaN)
    Fill existing missing (NaN) values, and any new element needed for
    successful Series alignment, with this value before computation.
    If data in both corresponding Series locations is missing
    the result will be missing.
level : int or name
    Broadcast across a level, matching Index values on the
    passed MultiIndex level.

Returns
-------
Series
    The result of the operation.

See Also
--------
Series.rtruediv

Examples
--------
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> a
a    1.0
b    1.0
c    1.0
d    NaN
dtype: float64
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> b
a    1.0
b    NaN
d    1.0
e    NaN
dtype: float64
>>> a.divide(b, fill_value=0)
a    1.0
b    inf
c    inf
d    0.0
e    NaN
dtype: float64

################################################################################
################################## Validation ##################################
################################################################################

3 Errors found:
	Parameters {axis} not documented
	Parameter "other" has no description
	Missing description for See Also "Series.rtruediv" reference

################################################################################
###################### Docstring (pandas.Series.floordiv) ######################
################################################################################

Return Integer division of series and other, element-wise (binary operator `floordiv`).

Equivalent to ``series // other``, but with support to substitute a fill_value for
missing data in one of the inputs.

Parameters
----------
other : Series or scalar value
fill_value : None or float value, default None (NaN)
    Fill existing missing (NaN) values, and any new element needed for
    successful Series alignment, with this value before computation.
    If data in both corresponding Series locations is missing
    the result will be missing.
level : int or name
    Broadcast across a level, matching Index values on the
    passed MultiIndex level.

Returns
-------
Series
    The result of the operation.

See Also
--------
Series.rfloordiv

Examples
--------
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> a
a    1.0
b    1.0
c    1.0
d    NaN
dtype: float64
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> b
a    1.0
b    NaN
d    1.0
e    NaN
dtype: float64
>>> a.floordiv(b, fill_value=0)
a    1.0
b    NaN
c    NaN
d    0.0
e    NaN
dtype: float64

################################################################################
################################## Validation ##################################
################################################################################

3 Errors found:
	Parameters {axis} not documented
	Parameter "other" has no description
	Missing description for See Also "Series.rfloordiv" reference
</pre>
</details>
<details><summary>Example ValueError from axis parameter</summary>
<pre>
>>> a = pd.Series([1, 1, 1, np.nan], index=['a', 'b', 'c', 'd'])
>>> b = pd.Series([1, np.nan, 1, np.nan], index=['a', 'b', 'd', 'e'])
>>> a.add(b)
a    2.0
b    NaN
c    NaN
d    NaN
e    NaN
dtype: float64
>>> a.add(b,axis=0)
a    2.0
b    NaN
c    NaN
d    NaN
e    NaN
dtype: float64
>>> a.add(b,axis=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/danlaw/Projects/pandas/pandas/core/ops.py", line 2044, in flex_wrapper
    self._get_axis_number(axis)
  File "/Users/danlaw/Projects/pandas/pandas/core/generic.py", line 361, in _get_axis_number
    .format(axis, type(cls)))
ValueError: No axis named 1 for object type <class 'type'>
</pre>
</details>

TomAugspurger commented on pandas/core/ops.py:
```diff
@@ -906,8 +1056,12 @@ def _make_flex_doc(op_name, typ):
-        doc = base_doc.format(desc=op_desc['desc'], op_name=op_name,
-                              equiv=equiv, reverse=op_desc['reverse'])
+        doc_no_examples = base_doc.format(desc=op_desc['desc'],
+            op_name=op_name, equiv=equiv, reverse=op_desc['reverse'])
```
I think this is what failed the build. Should be aligned with the open paren on the previous line, or move `desc=...` down here.

danielplawrence committed:
DOC: break out _make_flex_doc format arguments on seperate lines (#24589)

jreback commented:
@TomAugspurger

TomAugspurger commented:
Thanks @danielplawrence. The divmod failure is being worked on elsewhere, and documenting `axis` is out of scope for this PR.
